core/src/protocols_handler/mod.rs

```diff
@@ -134,28 +147,41 @@ pub trait ProtocolsHandler {
     fn inject_event(&mut self, event: Self::InEvent);
 
     /// Indicates to the handler that upgrading a substream to the given protocol has failed.
-    fn inject_dial_upgrade_error(&mut self, info: Self::OutboundOpenInfo, error: ProtocolsHandlerUpgrErr<<Self::OutboundProtocol as OutboundUpgrade<Self::Substream>>::Error>);
+    fn inject_dial_upgrade_error(
+        &mut self,
+        info: Self::OutboundOpenInfo,
+        error: ProtocolsHandlerUpgrErr<
+            <Self::OutboundProtocol as OutboundUpgrade<Self::Substream>>::Error
+        >
+    );
 
     /// Returns until when the connection should be kept alive.
     ///
-    /// If returns `Until`, that indicates that this connection may be closed and this handler
-    /// destroyed after the returned `Instant` has elapsed if they think that they will no longer
-    /// need the connection in the future. Returning `Forever` is equivalent to "infinite".
-    /// Returning `Now` is equivalent to `Until(Instant::now())`.
+    /// This method is called by the `Swarm` after each invocation of
+    /// [`ProtocolsHandler::poll`] to determine if the connection and the associated
+    /// `ProtocolsHandler`s should be kept alive and if so, for how long.
+    ///
+    /// Returning [`KeepAlive::Now`] indicates that the connection should be
+    /// closed and this handler destroyed immediately.
     ///
-    /// On the other hand, the return value is only an indication and doesn't mean that the user
-    /// will not close the connection.
+    /// Returning [`KeepAlive::Until`] indicates that the connection may be closed
+    /// and this handler destroyed after the specified `Instant`.
     ///
-    /// When multiple `ProtocolsHandler` are combined together, the largest `KeepAlive` should be
-    /// used.
+    /// Returning [`KeepAlive::Forever`] indicates that the connection should
+    /// always be kept alive. The connection will only be closed if
```

Comment:
Should "always" be kept alive is a bit misleading in my opinion:
- We may allow the user to forcibly close connections in the future.
- When `connection_keep_alive()` is called again, it can return a value other than `Forever`.

Reply:
Good point, that should be reworded.